pkg/controllers/provisioning/static/controller.go

```diff
@@ -0,0 +1,146 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package static
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/samber/lo"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	nodeutils "sigs.k8s.io/karpenter/pkg/utils/node"
+	"sigs.k8s.io/karpenter/pkg/utils/resources"
+
+	"sigs.k8s.io/karpenter/pkg/controllers/provisioning"
+	"sigs.k8s.io/karpenter/pkg/controllers/state"
+	"sigs.k8s.io/karpenter/pkg/events"
+	"sigs.k8s.io/karpenter/pkg/metrics"
+
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+
+	"sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/operator/injection"
+	nodepoolutils "sigs.k8s.io/karpenter/pkg/utils/nodepool"
+)
+
+type Controller struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+	cluster       *state.Cluster
+	ncProvisioner *provisioning.NCProvisioner
+}
+
+func NewController(kubeClient client.Client, cluster *state.Cluster, recorder events.Recorder, cloudProvider cloudprovider.CloudProvider) *Controller {
+	return &Controller{
+		kubeClient:    kubeClient,
+		cloudProvider: cloudProvider,
+		cluster:       cluster,
+		ncProvisioner: provisioning.NewNCProvisioner(kubeClient, recorder, cluster),
+	}
+}
+
+// Reconcile the resource
+func (c *Controller) Reconcile(ctx context.Context, np *v1.NodePool) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, "provisioning.static")
+	// TODO: This can be improved to not wait for cluster sync but we need to either rely on a manual
+	// update of NodePoolResources so that we track node count accurately, or handle getting NodeClaim count some other way
+	if !c.cluster.Synced(ctx) {
+		return reconcile.Result{RequeueAfter: time.Second}, nil
+	}
+	if !nodepoolutils.IsManaged(np, c.cloudProvider) {
+		return reconcile.Result{}, nil
+	}
+	if !np.StatusConditions().Root().IsTrue() {
+		return reconcile.Result{}, nil
+	}
+	if np.Spec.Replicas == nil {
+		return reconcile.Result{}, nil
+	}
+	nodes := c.cluster.NodePoolResourcesFor(np.Name)[resources.Node]
+	// Size down of replicas will be handled in disruption controller to drain nodes and delete NodeClaims
+	if nodes.Value() > lo.FromPtr(np.Spec.Replicas) {
```

Comment:
Because we are relying on internal state to handle the count of nodes here, I think there _may_ be some cases where the event filter may cause us to miss events here -- notably, if there are cases where we have scaled down but the cluster state hasn't seen it yet, we may not have an event that causes us to get requeued when the cluster state _does_ finally get updated -- it is unlikely, but it may still be worth a requeue of every minute or so to ensure that we don't accidentally miss an event when we think we are above our threshold